io/src/main/scala/fs2/io/net/Socket.scala

```diff
@@ -69,3 +76,139 @@ trait Socket[F[_]] {
+      readSemaphore: Semaphore[F],
+      writeSemaphore: Semaphore[F]
+  )(implicit F: Async[F])
+      extends Socket[F] {
+    private[this] var readBuffer: ByteBuffer = ByteBuffer.allocateDirect(8192)
+
+    private def getBufferForRead(size: Int): F[ByteBuffer] = F.delay {
+      if (readBuffer.capacity() < size)
+        readBuffer = ByteBuffer.allocateDirect(size)
+      else
+        (readBuffer: Buffer).limit(size)
+      readBuffer
+    }
+
+    /** Performs a single channel read operation in to the supplied buffer. */
+    private def readChunk(buffer: ByteBuffer): F[Int] =
+      F.async_[Int] { cb =>
+        ch.read(
+          buffer,
+          null,
+          new CompletionHandler[Integer, AnyRef] {
+            def completed(result: Integer, attachment: AnyRef): Unit =
+              cb(Right(result))
+            def failed(err: Throwable, attachment: AnyRef): Unit =
+              cb(Left(err))
+          }
+        )
+      }
+
+    /** Copies the contents of the supplied buffer to a `Chunk[Byte]` and clears the buffer contents. */
+    private def releaseBuffer(buffer: ByteBuffer): F[Chunk[Byte]] =
+      Async[F].delay {
+        val read = buffer.position()
+        val result =
+          if (read == 0) Chunk.empty
+          else {
+            val dest = new Array[Byte](read)
+            (buffer: Buffer).flip()
+            buffer.get(dest)
+            Chunk.array(dest)
+          }
+        (buffer: Buffer).clear()
+        result
+      }
+
+    def read(max: Int): F[Option[Chunk[Byte]]] =
+      readSemaphore.permit.use { _ =>
+        getBufferForRead(max).flatMap { buffer =>
+          readChunk(buffer).flatMap { read =>
+            if (read < 0) Async[F].pure(None)
+            else releaseBuffer(buffer).map(Some(_))
+          }
+        }
+      }
+
+    def reads(maxBytes: Int): Stream[F, Byte] =
+      Stream.eval(read(maxBytes)).flatMap {
+        case Some(bytes) =>
+          Stream.chunk(bytes) ++ reads(maxBytes)
+        case None => Stream.empty
+      }
+
+    def readN(max: Int): F[Option[Chunk[Byte]]] =
+      readSemaphore.permit.use { _ =>
+        getBufferForRead(max).flatMap { buffer =>
+          def go: F[Option[Chunk[Byte]]] =
+            readChunk(buffer).flatMap { readBytes =>
+              if (readBytes < 0 || buffer.position() >= max)
+                // read is done
+                releaseBuffer(buffer).map(Some(_))
+              else go
+            }
+          go
+        }
+      }
+
+    def write(bytes: Chunk[Byte]): F[Unit] = {
+      def go(buff: ByteBuffer): F[Unit] =
+        Async[F]
+          .async_[Unit] { cb =>
+            ch.write(
+              buff,
+              (),
+              new CompletionHandler[Integer, Unit] {
+                def completed(result: Integer, attachment: Unit): Unit =
+                  cb(Right(()))
+                def failed(err: Throwable, attachment: Unit): Unit =
+                  cb(Left(err))
+              }
+            )
```

Comment:
We cal also use the `CallBackHandler` here: 

```suggestion
            ch.write(buff, (), new CallBackHandler[Integer, Unit](cb))
```